Source/Core/VideoBackends/Vulkan/main.cpp

```diff
@@ -89,11 +98,51 @@ static bool ShouldEnableDebugReports(bool enable_validation_layers)
   return enable_validation_layers || IsHostGPULoggingEnabled();
 }
 
+// Helpers to manage the connection to the X server.
+bool VideoBackend::OpenDisplayConnection()
+{
+#if defined(VK_USE_PLATFORM_XCB_KHR)
+  // We could probably use a xcb_connection_t here, however since we get passed an XLib window
+  // handle, not a xcb_window_t handle, to keep things consistent we will use an XLib connection.
+  m_display_handle = XOpenDisplay(nullptr);
+  if (!m_display_handle)
+  {
+    PanicAlert("Failed to open X display.");
+    return false;
+  }
+#endif
+
+  return true;
+}
+
+void VideoBackend::CloseDisplayConnection()
+{
+#if defined(VK_USE_PLATFORM_XCB_KHR)
+  if (m_display_handle)
+  {
+    XCloseDisplay(reinterpret_cast<Display*>(m_display_handle));
+    m_display_handle = nullptr;
+  }
+#endif
+}
+
 bool VideoBackend::Initialize(void* window_handle)
 {
+  // If we are running headless, we don't need the surface extensions or a display connection.
+  bool enable_surface = window_handle != nullptr;
+
+  // Ensure we have a connection to the display server (if any) before continuing.
+  if (enable_surface && !OpenDisplayConnection())
+    return false;
+
+  // Dynamically load vulkan-1.dll/libvulkan.so.
   if (!LoadVulkanLibrary())
   {
-    PanicAlert("Failed to load Vulkan library.");
+    PanicAlert("Failed to load Vulkan library.\n\n"
+               "This may be due to the loader not being installed on your system.\n\n"
+               "The loader is usually provided with your display driver on Windows, "
+               "or in a package named 'libvulkan1' or 'vulkan-icd-loader' on Linux.");
+    CloseDisplayConnection();
```

Comment:
That does sound like a cleaner way of doing things. However, the "VideoBackend" classes are created at startup, and then "selected" with this method: https://github.com/dolphin-emu/dolphin/blob/master/Source/Core/VideoCommon/VideoBackendBase.cpp#L76

IMO a better way would be for the backends to provide factory methods instead, and contain everything within the interface returned. Therefore, "destroying" the backend would just be a matter of releasing this pointer.

The fact that there are two "cleanup" methods does complicate things (one runs on the video thread, one on the main thread), but this is probably indicative of a larger design issue.